ApiReview/API.Polly.Core/NoDocs/Polly/ResilienceProperties.cs

```diff
@@ -0,0 +1,15 @@
+// Assembly 'Polly.Core'
+
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+
+namespace Polly;
+
+public sealed class ResilienceProperties
```

Comment:
Andrey:
can we mimic the behaviour of `IFeatureCollection` here by changing the API to use the `TValue` type as a key? this eliminates the need to generate custom names for properties and have the artificial `PropertyKey` type and fosters the everything is strong typed approach

```cs
public class ResilienceProperties
{
    void Set<TValue>(TValue? value);
    TValue? Get<TValue>();
}
```

Martin:
One common use-case of V7 is attaching primitive types (see #890 as an example). While in V8 they could use something like `properties.Set<int>(3)` or `properties.Set<MyIntWrapper>()` it could lead to ambiguity or more more complicated usage.

Reply:
can you elaborate on this a bit more
- the current approach implies that a caller must create an instance of PropertyKey to hold the key
- the suggested approach allows a caller to use custom types as well - making the values strong typed - but also primitives as you pointed out above

Reply:
> the current approach implies that a caller must create an instance of PropertyKey to hold the key

Correct, we want to avoid magic strings or value casting. Key encapsulates both type and key to the value. 

> the suggested approach allows a caller to use custom types as well - making the values strong typed - but also primitives as you pointed out above

Yes, there might be some performance penalty if you want to store `string` for example as extra wrapper will be required to store the value. 

Another downside is unfamiliarity for Polly V7 users. Migrating from `Context` to `ResilienceProperties` is relatively straightforward. You don't even have to hold the key, you can just create it inline.

This API is also familiar for users that work with HttpClient:
- https://source.dot.net/#System.Net.Http/System/Net/Http/HttpRequestOptionsKey.cs,12

Or with RateLimiting:
- https://source.dot.net/#System.Threading.RateLimiting/System/Threading/RateLimiting/MetadataName.T.cs,12

Reply:
on the contrary, the typed approach must be familiar to all ASP.NET users

Reply:
I don't follow the suggestion here. Can you give some examples in code of usage with the current code and with the proposed change?

IFeatureCollection is typically 1:1 is it not, with there only ever being a single instance of a given type representing a concept? It seems to me that this suggestion would make it harder to attach multiple instances of a specific type to the context, such as strings where you might want multiple types of IDs associated with it.

Reply:
ASP.NET Core's typed approach requires users to recognize that every entity passed through a pipeline via the context object represents a contract between two or more components, and as such, it'd better be well-defined, provide encapsulation, be future-proof, and so on. 

imo, the current approach, on the other hand, encourages the usage of the primitive obsession pattern by focusing solely on the keys rather than te entities.

> you might want multiple types of IDs

every ID in this example has some role, some rules and agreements around it, may be every ID must be encoded according to some format, there is a set of operations that can be performed or even things like what to do when it isn't presented. all these aspects can't be transferred over with a mere string entity. instead, the approach chosen by the ASP.NET Core team forces one to use the OOP principles to encode the contract by creating a strong type for the entities and implementing the contract within the new type. For example

```cs
public class MyIDs
{
     string MyFirstId { get; init; }
     string? MySeconfId { get; init; }
}

// publish
var ids = new MyIDs(..);
ResilienceProperties.Set(ids);

// retrieve
var myIds= ResilienceProperties.Get<MyIDs>();
```

if the ids need to be modifiable, one can reflect that via the API of the type. if some of the ids are optiona, make the properties nullable. if the ids aren't simple arbitrary strings, add the validation to the ctor.

Reply:
Keeping as is.

Reply:
fair enough